repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/AssetPreProcessor.java

```diff
@@ -0,0 +1,116 @@
+                                  EntityMutations.EntityOperation operation) throws AtlasBaseException {
+        if (LOG.isDebugEnabled()) {
+            LOG.debug("AssetPreProcessor.processAttributes: pre processing {}, {}",
+                    entityStruct.getAttribute(QUALIFIED_NAME), operation);
+        }
+        this.context = context;
+
+        AtlasEntity entity = (AtlasEntity) entityStruct;
+
+        AtlasVertex vertex = context.getVertex(entity.getGuid());
+
+        switch (operation) {
+            case CREATE:
+                processCreateAsset(entity, vertex);
+                break;
+            case UPDATE:
+                processUpdateAsset(entity, vertex);
+                break;
+        }
+    }
+
+    private void processCreateAsset(AtlasEntity entity, AtlasVertex vertex) throws AtlasBaseException {
+        AtlasPerfMetrics.MetricRecorder metricRecorder = RequestContext.get().startMetricRecord("processCreateAsset");
+
+        if(entity.hasAttribute(DOMAIN_GUIDS)) {
+          validateDomainAssetLinks(entity);
+          AtlasEntityHeader sourceEntity = new AtlasEntityHeader(entity);
+          isAuthorized(sourceEntity);
+        }
+
+        RequestContext.get().endMetricRecord(metricRecorder);
+    }
+
+
+    private void processUpdateAsset(AtlasEntity entity, AtlasVertex vertex) throws AtlasBaseException {
+        AtlasPerfMetrics.MetricRecorder metricRecorder = RequestContext.get().startMetricRecord("processUpdateAsset");
+
+        if(entity.hasAttribute(DOMAIN_GUIDS)) {
+            validateDomainAssetLinks(entity);
+            AtlasEntityHeader sourceEntity = new AtlasEntityHeader(entity);
+            isAuthorized(sourceEntity);
+        }
+
+        RequestContext.get().endMetricRecord(metricRecorder);
+
+    }
+
+    private void validateDomainAssetLinks(AtlasEntity entity) throws AtlasBaseException {
+        List<String> domainGuids = ( List<String>) entity.getAttribute(DOMAIN_GUIDS);
+
+        if(domainGuids.size() > 1) {
+            throw new AtlasBaseException(AtlasErrorCode.INVALID_PARAMETERS, "Asset can be linked to only one domain");
+        }
+
+        if(CollectionUtils.isNotEmpty(domainGuids)) {
+            for(String domainGuid : domainGuids) {
+                AtlasVertex domainVertex = entityRetriever.getEntityVertex(domainGuid);
+                  if(domainVertex == null) {
+                        throw new AtlasBaseException(AtlasErrorCode.INSTANCE_GUID_NOT_FOUND, domainGuid);
+                  }
+                  else{
+                      LOG.info("Domain vertex found for guid: {}", domainGuid);
+                  }
+            }
+        }
+    }
+
+    private void isAuthorized(AtlasEntityHeader sourceEntity) throws AtlasBaseException {
+
+        // source -> CREATE + UPDATE + READ
+        AtlasAuthorizationUtils.verifyAccess(new AtlasEntityAccessRequest(typeRegistry, AtlasPrivilege.ENTITY_UPDATE, sourceEntity),
```

Comment:
Based on the operation (create or update) we can avoid one of the auth checks here
E.g. In case the Asset is being updated, avoid ENTITY_UPDATE check as it will be anyway happen before even executing pre-processors

Reply:
Comment says CREATE + UPDATE + READ, but only 2 are here
CREATE is not needed?

Reply:
Yes the decision was changed, I will change the comment

Reply:
Still ENTITY_UPDATE auth check is redundant here in case operation is UPDATE asset, in case execution reaches here, it is already performed for the asset, you can remove this check